src/core/dynamic_batch_scheduler.cc

```diff
@@ -153,210 +153,202 @@ DynamicBatchScheduler::Enqueue(std::unique_ptr<InferenceRequest>& request)
+    }
+  }
+
+  NewPayload();
+
+  while (!scheduler_thread_exit_.load()) {
+    NVTX_RANGE(nvtx_, "DynamicBatcher " + model_->Name());
+
+    std::shared_ptr<std::vector<std::deque<std::unique_ptr<InferenceRequest>>>>
+        rejected_requests;
+    uint64_t wait_microseconds = 0;
+
+    auto payload_state = curr_payload_->GetState();
+    if (payload_saturated_ ||
+        payload_state == RateLimiter::Payload::State::EXECUTING ||
+        payload_state == RateLimiter::Payload::State::RELEASED) {
+      NewPayload();
+      next_preferred_batch_size_ = 0;
+      required_equal_inputs_.clear();
+    }
+
+    // Hold the lock for as short a time as possible.
+    {
+      std::unique_lock<std::mutex> lock(mu_);
+      auto requests_available = [this]() {
+        return (!queue_.Empty()) || scheduler_thread_exit_.load();
+      };
+      cv_.wait(lock, requests_available);
+
+      if (delay_cnt > 0) {
+        // Debugging/testing... wait until queue contains 'delay_cnt'
+        // items...
+        wait_microseconds = 10 * 1000;
+        if (queue_.Size() >= delay_cnt) {
+          delay_cnt = 0;
+        }
+        LOG_VERBOSE(1) << "Delaying batcher thread " << model_->Name()
+                       << " until " << delay_cnt
+                       << " queued requests, current total = " << queue_.Size();
+      } else if (queue_.Empty()) {
+        wait_microseconds = 0;
+      } else {
+        if (payload_saturated_) {
+          continue;
+        }
+        auto wait_for_slots = [this]() {
+          return model_->Server()->GetRateLimiter()->PayloadSlotAvailable(
+              model_);
+        };
+        slot_cv_.wait(lock, wait_for_slots);
+        {
+          std::lock_guard<std::mutex> exec_lock(
+              *(curr_payload_->GetExecMutex()));
+
+          auto payload_state = curr_payload_->GetState();
+          if (payload_state == RateLimiter::Payload::State::EXECUTING ||
+              payload_state == RateLimiter::Payload::State::RELEASED) {
+            continue;
+          }
+
+          // Use dynamic batching to get request(s) to execute.
+          wait_microseconds = GetDynamicBatch();
+
+          // Get requests that are rejected from searching dynamic batch.
+          queue_.ReleaseRejectedRequests(&rejected_requests);
+
+          // Extract batch only if there is pending batch
+          auto pending_batch_queue_cnt = queue_.PendingBatchCount();
+          if ((wait_microseconds == 0) && (pending_batch_queue_cnt != 0)) {
+            curr_payload_->ReserveRequests(pending_batch_queue_cnt);
+            for (size_t idx = 0; idx < pending_batch_queue_cnt; ++idx) {
+              std::unique_ptr<InferenceRequest> request;
+              auto status = queue_.Dequeue(&request);
+              if (status.IsOk()) {
+                if (preserve_ordering_) {
+                  DelegateResponse(request);
+                }
+                curr_payload_->AddRequest(std::move(request));
```

Comment:
It may be more clear to have a SetState() function and calls it at the end of adding pending requests